### DISCUSSION TITLE
Root loader + shouldReload doesn't seem to work properly

### DISCUSSION THREAD

**Discussion by sturatcliffe** [remix-run/remix#1425] [Q&A] [answered]
Hey,

I'm not sure if I'm doing something wrong, misunderstanding how loaders/shouldReload are supposed to work, or if there's a bug around ```unstable_shouldReload``` / ```ShouldReloadFunction```.

In my app, every page besides SignIn / Register require the user to be authenticated. So rather than have to remember to auth check in the loader of every page, I'm trying to get a loader in ```root.tsx``` to do all of the auth checking for every page in the app. Essentially all this does is try and read the logged in user from the session, and if the current URL is anything except ```/signin``` or ```/register```, redirect to the login page if they aren't logged in.

However, I can't get the root loader to reload on every transition, which means when the users session expires they can keep navigating around until force refreshing the entire page. It also means that when they hit the sign out link, the UI doesn't update properly as the navbar is rendered in ```root.tsx```, and should only show if the user is logged in.

I've put together a massively simplified example here: https://github.com/sturatcliffe/remix-demo - with a working demo hosted here: https://remix-demo.ratcliffe.io/

After hitting the sign in button, the session cookie is set with a 10 second expiration, after which you can keep flicking between the home and profile pages until you refresh the browser or click on the same link as the page currently being viewed.

Any suggestions on how to handle this scenario? I know I can fix the UI issues by nesting all authenticated pages under a directory prefixed with double underscore, e.g. ```app/routes/__app/index.tsx``` and ```app/routes/__app/profile.tsx``` then move the navbar display out of ```root.tsx``` and into a layout route for the ```__app``` directory, but it feels a bit OTT for a simple scenario like this. And it doesn't fix the loader issue, even if I move the auth check out of the root loader and into a layout loader, it still doesn't reload on every transition.

Thanks in advance!

**Comment by chikamichi** [comment]
Same issue here.

I’m wondering whether session & session checking should be the reponsibility of the underlying server (eg. Express) rather than the app itself.

**Comment by kiliman** [comment, marked as answer]
I think you misunderstand how loaders work. First of all, each loader is independent. On initial page request, Remix will call the loaders of all matching routes (from root to leaf) in parallel. So results from one loader won't be available in another.

On subsequent navigations, Remix will only call loaders of changed routes. That is if you are going from `/parent/child-a` => `/parent/child-b`, only the loader for `child-b` will be called since they both share the common parent routes: `root` and `parent`.

That is why you're seeing the behavior you are. The `root` loader will never be called again during normal navigation.

It's only when you POST to an action that Remix will _revalidate_ the data by calling *all* the matching loaders in parallel from the client. So if you were to POST from `/parent/child-a`, it would first call the action in `child-a`, then upon return, it will call the loaders of: `root`, `parent` and `child-a` in parallel.

The `unstable_shouldReload` export is so that you can opt out of the data revalidation for a specific route. For example, if the `root` loader only contains static data that will never change, so doesn't need to be reloaded.

Currently Remix doesn't have the concept of middleware, but they are considering adding support for pre/post request handlers. In the meantime, you either need to add your user validation in *every* loader, or you can add the check in your server adapter (e.g., Express middleware) before handing the request to Remix.

**Reply by ratasorin** [reply]
So to answer the question "How can one properly log out a user?", you would make the `log out` button a submit button that triggers an `action`. Thus, the root loader would be triggered as well, and the session will have to be verified! (unless `unstable_shouldReload` is set to `false`). Right?
And for safe navigation we could use `<a>` instead of `<Links>` to also trigger the root! We could have the outer most navigation as `<a>` tags that would make sure the user is still authenticated and combine it with route specific `<Links>` that would give the website an SPA feel.

**Reply by sergiodxa** [reply]
For safe navigation you should check for auth on every loader instead of only on the root and giving up on client-side navigation for the outer most navigation.

Each loader is an individual API endpoint, and in a protected API endpoint you would check for auth on every request, here's the same.

**Reply by barbinbrad** [reply]
Do you have any recommendations for code style? Ideally, we'd write very minimal code to do this. But nothing great comes to mind because of the conventions.

**Reply by sturatcliffe** [reply]
@barbinbrad I've started using the same pattern as they're using in the Remix Stacks examples, e.g. here: https://github.com/remix-run/blues-stack/blob/main/app/routes/notes/%24noteId.tsx

The first line in every loader / action which requires auth, I do something like this:

```
const userId = await requireUserId(request);
```

It's repetitive for sure, and hopefully in the future we may get some kind of middleware feature which helps in this regard. However, I've actually grown to quite like it this way as you are very explicit with the requirements of each page, and there is no "magic" hiding the authentication checking away somewhere that isn't obvious to new developers who may look at your code.

Hope this helps.

**Reply by kiliman** [reply]
The way I do it is to add a check for the auth cookie in my Express entry point. If cookie not present, I redirect to /login, otherwise forward request to Remix.

I have a whitelist of routes that are "public" like /login, /register, etc.

This way ALL routes are protected by default, and I have to specifically opt-out of it. This way I won't forget to protect a new route.

I also use `requireUserId()` anywhere I explicitly need a user, to confirm the auth cookie is valid.

https://github.com/kiliman/express-auth-example/blob/4aed9f0bfeb85d7cc73d4e8232cbaf9cd1a1d1c3/server/index.js#L63-88

**Reply by barbinbrad** [reply]
> https://github.com/kiliman/express-auth-example/blob/4aed9f0bfeb85d7cc73d4e8232cbaf9cd1a1d1c3/server/index.js#L63-88

Super interesting! This is definitely less code. 

Is this limited to express-type setups or could this same thing be done in a cloudflare worker, for example?

**Reply by sergiodxa** [reply]
You can do it with Cloudflare too because you can control when you pass the request to Remix to handle it.
